rpcs3/rpcs3qt/gs_frame.cpp

```diff
@@ -359,6 +359,12 @@ bool gs_frame::get_mouse_lock_state()
 
 void gs_frame::close()
 {
+	if (m_is_closing.exchange(true))
```

Comment:
What happens if close is called after the frame has been deleted? m_is_closing will not necessarily contain anything useful then. Is it possible to fix the underlying problem?

Reply:
😿

Reply:
In theory this would be an issue.
But the current possibilities are this:

A: Stop emulation from outside the gs_frame:
- The still existing gs_frame is closed by rsx_thread
- delete gs_frame

B: Stop emulation from inside the gs_frame:
-  close(): // The rest of the stuff below happens blocking in the CallFromMainThread
-  Emu.Stop()
   -  recursive call to still existing gs_frame::close() from rsx_thread
   -  early out
-  delete gs_frame

Reply:
We're in a bit of a dilemma here:
If I call a non blocking Gracefulshutdown instead of closing the gs_frame internally,
then the game window will not respond while the game is terminating.
It would be weird if you click X and the game doesn't stop.
The only possible way to go here would be to display the "Please wait while the game is shutting down" message via native overlays. But the current rsx implementation is not mature enough to handle this, since there is no visual rsx shutdown period - which is the reason why the "shutdown" dialog is only working while we close the game during initial cache compilation.

Reply:
We can hide the gsframe while we stop things internally. Ideally, whoever spawned gsframe is responsible for destroying it.
Also, why is the call to close gsframe recursive?

Reply:
GracefulExit in close is blocking to keep the gsframe open while the shutdown message is displayed, this was a usability decision, because it is strange to have to wait for seconds after the frame is closed.
But stopping emulation will inevitably try to close the game window, therefore it is recursive.

I'll just replace the internal close with an async call to GracefulShutdown. Then the only way to close the window is via rsx_thread.
It seems to close quickly enough not to create a weird feeling when clicking X.